app/controllers/users_controller.rb

```diff
@@ -92,6 +92,32 @@ def teams
   end
   helper_method :teams
 
+  def opt_in_params
+    params.require(:user).permit(
+      :opted_in_newsletter_at,
+      :opted_in_announcements_at,
+      :opted_in_marketing_announcements_at,
+      :opted_in_surveys_at,
+      :opted_in_sponsorships_at,
+      :opted_in_applications_open_at
+    ).to_h.map do |opt_in, value|
+      next if opted_in?(opt_in_selected?(value), opt_in)
+      [opt_in, opted_in_time(value)]
+    end.to_h
```

Comment:
This is not pretty at all D: 

Any ideas how I could improve this, please?

Reply:
Sorry for coming back to this so late @ramonh 🙇 

I don't know the exact circumstances for this feature, but I'll just quickly dump my brain here on how I'd have implemented it if it would have been in this very second 😉 (disclaimer: this is all just written out of memory, copy-pasting will probably 💥)

You could have a GDPR model concern (bad name, but just for an example):
```ruby
module GDPR
  extend ActiveSupport::Concern

  included do
    def opted_in_newsletter_at=(value)
      super(nil) unless value # should be set to nil for falsey values
      return value if value && self.opted_in_newsletter # don't do anything if already set
      super(Time.current)
    end

    # other methods similar
  end
end
```

If you then include this into the user model, you can just use ordinary checkboxes in the form and pass the params through as they are, without some `to_h` etc. magic in the controller.

One should probably dry this up, think about edge-cases, etc. But just as an example on how one could do it 😉

Reply:
Wow, this is a cool idea! Thank you @klappradla! I'll look into this

Reply:
Hello! 👋

What I've done in the past is something similar to what this gem does: https://github.com/SonderCollection/timestamp_as_boolean

So, something along the lines of:

```ruby
  def opted_in_newsletter
    opted_in_newsletter_at.present?
  end
  alias opted_in_newsletter? opted_in_newsletter

  def opted_in_newsletter=(checkbox_value)
    accepted = ActiveRecord::Type::Boolean.new.cast(checkbox_value)
    if accepted
      self.opted_in_newsletter_at ||= Time.now
    else
      self.opted_in_newsletter_at = nil
    end
  end
```

Then you could use `opted_in_newsletter` in the checkboxes.

There may be other more known gems that do this - this one was what I found after a quick google search. I like the idea of making this a concern 👍 And it should be DRY'd too.

This is just to give you more options, if you need them - I like @klappradla's solution too 😃

Reply:
Cool! Thanks for the tip, @pgaspar. 

I'll weigh out both options. 

Thanks!